proposals/improved-hash-module.md

```diff
@@ -0,0 +1,134 @@
+
+## Example
+
+```mojo
+@value
+struct Person(Hashable):
+    var name: String
+    var age: UInt8
+    var friends_names: List[String]
+
+    fn __hash__(self) -> Int:
+        var hashes = List[Int]()
+        hashes.append(hash(self.name))
+        hashes.append(hash(self.age))
+        for friend in self.friends_names:
+            hashes.append(hash(friend[]))
+        
+        # How to combine a hash of hashes ???
+```
+As you can see above we, computed hashes for all of the struct fields, but we are uncertain how to combine those values in a way which produces a good (non compromised) hash value. Python (docs)[https://docs.python.org/3/reference/datamodel.html#object.__hash__] suggest to pack fileds into a tuple and hash the tuple, but this is not possible at this point in time in Mojo.
+
+# Proposal
+
+In order to improve the hash module and address the flaws of the `Hashable` trait, we need to apply following steps.
+
+## Introduce a `Hasher` trait
+
+By introducing a `Hasher` trait we define an abstraction for the hashing algorithm itself, which allows streaming creation of the hash value. Bellow is a possible API of the `Hashable` trait.
+
+```mojo
+trait Hasher:
+    fn __init__(inout self):
+        ...
+    fn update[T: DType](inout self, value: SIMD[T, 1]):
+        ...
+    fn update(inout self, pointer: DTypePointer[DType.uint8], length: Int):
+        ...
+    fn finish(owned self) -> UInt64:
+        ...
+```
+
+## Implement a default `Hasher` in standard library
+
+The standard library should provide a default `Hasher` implementation, but it would be possible for the developers to implement, or choose other hash algorithms, if they better fit their use case.
+
+Bellow you can see a dummy implementation of a `DefaultHasher`
+
+```mojo
+struct DefaultHasher(Hasher):
+    var hash: UInt64
+
+    fn __init__(inout self):
+        self.hash = 42
+    fn update[T: DType](inout self, value: SIMD[T, 1]):
+        ...
+    fn update(inout self, pointer: DTypePointer[DType.uint8], length: Int):
+        ...
+    fn finish(owned self) -> UInt64:
+        return self.hash
+```
+
+## Redesign the `Hashable` trait to follow the data flow principles
+
+Given the `Hasher` trait, we can define the `Hashable` trait to adopt the data flow paradigm instead of call return.
+
+```mojo
+trait Hashable:
+    fn hash_with[H: Hasher](self, inout hasher: H):
+        ...
+```
+
+## Example for `Hashable` struct implementation
+
+The implementation of `Hashable`, where all the fields are `Hashable` is trivial and could be easily synthesized by the compiler. But even if not, the data flow API guides the developer toward a very simple solution
+
+```mojo
+@value
+struct Person(Hashable):
+    var name: String
+    var age: UInt8
+    var friends_names: List[String]
+
+    fn hash_with[H: Hasher](self, inout hasher: H):
```

Comment:
Python and Mojo both use dunder `__something__` to signify that methods have a special meaning in the stdlib and the language. It avoids naming conflicts with user methods in general. I don't have a strong opinion about this, I'm just pointing this out.
This method in this proposal shouldn't be a second class citizen compared to `__hash__`.

Reply:
I agree, I created a small POC, where I changed the method names. https://gist.github.com/mzaks/aa66c831dc5e177c2322d5088aac76aa

If it is ok with everyone I will change the snippets in the proposal and reference the gist at the end.